### DISCUSSION TITLE
Updating Project Packages and Migrating from Depricated Packages

### DISCUSSION THREAD

**Discussion by thabaum** [oqtane/oqtane.framework#2798] [Q&A]
The following Oqtane Framework installed packages are deprecated:

System.ComponentModel.Annotations
Microsoft.AspNetCore.Mvc.ViewFeatures
Microsoft.AspNetCore.Localization

Most Microsoft packages need updated to 6.0.16 would be the final version for .NET 6 LTS in Oqtane 3.4 

Other packages support both 6 and 7 .NET Core.

With plans to make a .NET 7 release, that release will use the 7.0.5+ versions from Microsoft.

What is the plan to get Oqtane up-to-date so it doesn't smell of old code.  I understand the localization will need rebuilt to use the latest code trends with `Microsoft.Extensions.Localization` package and `IStringLocalizer`.

With Oqtane version 4 if anything needs to break to get up to speed I think this is the time to do it.  We may need to leave an old deprecated version but we should implement the new latest code trends to keep things tasty.

I was thinking about updating packages for the final .NET 6 version but we may just skip that and move to the .NET 7 updates?

I can attempt updates on the packages one PR at a time to get familiar in this area so we can keep things current going forward.

Thank you.

**Comment by sbwalker** [comment]
The Oqtane Philosophy (https://www.oqtane.org/blog/!/20/oqtane-philosophy) stresses the importance of backward compatibility. As such Oqtane 4.0 is not expected to have any breaking changes.

I am going to ignore the "smell of old code" comment as I assume it was either based on ignorance or more likely, just a poor choice of words.

**Comment by thabaum** [comment]
poor choice of words my ignorance is horrible a lot on my mind as the terminology used was not to offend a cutting edge technology like the Oqtane Framework, it was just referencing the deprecated packages that could give a bad taste such as not using Microsoft.Extensions.Localization instead (keep the existing deprecated but add the new way which is called in updated Oqtane modules but developers would need to update over time to keep up at some point maybe?! or is what is in Oqtane there for life?) and the last version of packages released 6.0.16 packages instead of 6.0.3.  It was a figure of speech nothing personal.  And so with that being said I am putting my feelers out here to see where I can or can't help.  I am pushing it's limits if you can't tell to see where it breaks for me so I can gauge what types of projects I may or may not wish to use with Oqtane Framework. 

Thanks for answering my question.  I will be sure to read all the blogs again to refresh all of this.

**Comment by sbwalker** [comment]
@hishamco perhaps you have experience with the Localization topic mentioned above? Basically, the Microsoft.AspNetCore.Localization Nuget package is flagged as deprecated by Microsoft (https://www.nuget.org/packages/Microsoft.AspNetCore.Localization). The Microsoft.Extensions.Localization package handles the vast majority of Localization use cases... and Oqtane already uses it extensively for localizing components (ie. using IStringLocalizer, etc...). The one specific area where Microsoft.AspNetCore.Localization appears to be used is for the CookieRequestCultureProvider. Is there an alternative to CookieRequestCultureProvider which should be used in newer versions of .NET?

**Reply by hishamco** [reply]
Seems many packages have been deprecated around two weeks ago according to the https://github.com/dotnet/announcements/issues/217 I knew many frameworks such as Oqtane & Orchard Core heavily use such packages

> Is there an alternative to CookieRequestCultureProvider which should be used in newer versions of .NET?

Frankly, I don't know, I just pushed 2 PRs for .NET 8.0 but they will affect `Microsoft.Extensions.Localization` package, let me check and come back to you

BTW is there an issue related to localization occurring? or just a question about why the packages were flagged as `Deprecated`?

**Reply by hishamco** [reply]
Quote from dotnet/core repo

> To provide developers with better package support status information, the .NET team will start using the [NuGet Package Deprecation feature](https://docs.microsoft.com/en-us/nuget/nuget-org/deprecate-packages). Deprecated packages remain visible on nuget.org, to clients such as Visual Studio and the .NET CLI, and can be restored by projects. In addition to updating existing packages, we will mark packages as appropriate when new versions are released so that developers can access up-to-date information on the packages they use.

**Comment by sbwalker** [comment]
@hishamco there are no issues related to localization - it is working well. I was more curious if it would be possible to remove the dependency on Microsoft.AspNetCore.Localization - but that would only be possible if there was an alternative for managing CookieRequestCultureProvider - and it sounds like you are not aware of an alternative currently. So I guess the best approach is to continue to use Microsoft.AspNetCore.Localization (even if Microsoft has flagged it as deprecated - which seems like a premature decision on their part since they have not provided a path forward).

**Reply by thabaum** [reply]
@sbwalker can you point me out to anything relating to `Microsoft.AspNetCore.Localization`

I may be able to come up with a solution to remove this dependency from the Oqtane Framework.  I just want to be sure I am working in the right areas covering all of it's use cases and then test on top of the net8 branch.  I will target any pull request on this branch once I get something produced that I feel may be acceptable and provides the same `CookieRequestCultureProvider` functionality required to remove this dependency.

From a search it looks like it is being used in the following files:

Oqtane.Client:
`\Modules\Admin\Languages\Add.razor`
`Program.cs`
`\Themes\Controls\Theme\LanguageSwitcher.razor`

Oqtane.Server:
`\Pages\_Host.cshtml.cs`

If this is all then I will try to create a middleware, culture provider, database-driven culture or settings type of solution that can handle this logic and remove the deprecated dependency if this is all that package is used for.  I will also do some asking around to see what feedback I can get.

If this pretty much wraps it up and I am on the right track no need for response I will work on this when motivated with time.

**Comment by hishamco** [comment]
> So I guess the best approach is to continue to use Microsoft.AspNetCore.Localization (even if Microsoft has flagged it as deprecated - which seems like a premature decision on their part since they have not provided a path forward).

Right, seem the flag was added to indicate there's no further maintenance place for such packages

**Comment by thabaum** [comment]
I asked on ASP.NET discussions about the deprecated package issue and the response from an MVP was:

> The API isn't deprecated, but the 2.2.0 NuGet package is because ASP.NET Core is out of support and since .NET Core 3.0 the API is part of the ASP.NET Core Shared Framework.
> You just need to [use a framework reference instead of a package reference](https://learn.microsoft.com/aspnet/core/fundamentals/target-aspnetcore#use-the-aspnet-core-shared-framework).

So we can remove this package dependency from Oqtane Framework by entering a "Framework Reference" like shown below:
```

<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp3.1</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
  </ItemGroup>

</Project>

```
[remove obsolete package references](
https://learn.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-8.0&tabs=visual-studio#remove-obsolete-package-references)

This should help keep Oqtane Framework project smelling fresh like it is when package dependencies are being reviewed.

**Reply by sbwalker** [reply]
That guidance does not make sense. The target framework in Oqtane is already net7 and soon to be net8 - which is already higher than netcoreapp3.1

**Reply by thabaum** [reply]
From my understanding this would be an "example" of how to edit the project file to remove an obsolete dependency package and include it as a package reference:

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.App"/>
    <PackageReference Include="Microsoft.AspNetCore.Localization" Version="2.2.0" PrivateAssets="All" />
  </ItemGroup>
</Project>
```

I think something like this is what you would include, then you can just remove the package dependency as it comes included with ASP .NET since they no longer plan to update it?


[Remove Obsolete Package References](https://learn.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-8.0&tabs=visual-studio#remove-obsolete-package-references
)

`<PackageReference Include="Microsoft.AspNetCore.Localization" Version="2.2.0" PrivateAssets="All" />`

This should be the line to focus including to remove the package from directions I have read.

I can try it out tomorrow I have a few things on my plate today and test around to see if I can make it build adding something similar to Oqtane project that relies on it.

From my understanding you include a "Package Reference" and then remove the "Package Dependency".

I just copied from the Microsoft Docs, I did not edit it for the Framework since I had not tested anything to confirm how exactly to place it in the file.

I do not undestand fully `PrivateAssets="All"` and if we can just include what Oqtane is needed to limit accessing the rest of the package that is supported in the newer localization package provided by Microsoft.  I would limit it as much as possible.

**Reply by thabaum** [reply]
Yes, now reviewing the information more, it just installs the packages having "package references" so you don't accomplish anything by doing this.

There are actually 3 obsolete packages being used by Oqtane. Removing the Microsoft.AspNetCore.Localization also removes the following

```
Microsoft.AspNetCore.Localization
Microsoft.AspNetCore.Http
Micrsosoft.Net.Http.Headers

```

I have not created an issue for Oqtane so this is just a "discussion" no need to get too worried about it as .NET 8 updates are the focus.

For reference these are the errors when these packages are removed:

![image](https://github.com/oqtane/oqtane.framework/assets/13577556/669da114-c6c8-43e4-aca9-b34422fc40a6)

If/When it gets worked on to remove the deprecated packages if possible from Oqtane this just gives a little more insight on what all will be involved.

**Reply by thabaum** [reply]
From my understanding if you include 

```
<ItemGroup>
  <FrameworkReference Include="Microsoft.AspNetCore.App" />
</ItemGroup>

```

In the server project these obsolete packages will build without a package reference.

Problem is Oqtane uses these in it's Client project which does not support WASM WebAssembly Blazor Client to use these dependencies in this way.

[https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/6.0/microsoft-aspnetcore-http-features-package-split](https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/6.0/microsoft-aspnetcore-http-features-package-split)

is another example

And here is a thread with the same issue... no solution.

[https://stackoverflow.com/questions/76173982/migrating-the-deprecated-microsoft-aspnetcore-http-abstractions-nuget-package](https://stackoverflow.com/questions/76173982/migrating-the-deprecated-microsoft-aspnetcore-http-abstractions-nuget-package)

If this logic moves to the server project we could use the solution and I will test a dependency like it in the server project and report back.

Still these are obsolete and worry me for vulnerabilities as no one is looking out for them anymore and moved on past these.  This does provide bit more research and understanding about the issue.

**Comment by sbwalker** [comment]
@thabaum I am not currently concerned about these packages as they are so widely used by so many applications. I assume there's no further maintenance expected for these packages because they are very mature.

**Reply by thabaum** [reply]
@sbwalker  I have put a bug in the ear of .NET Core project to get their solution to work with WASM if it is "mature" and not something that can introduce security issues.  I want to hear it directly from Microsoft there are not any issues security wise to worry about when using these packages.  Thanks for the feedback, sorry to keep digging in here but if the packages are great and simply are not needing updates they should be available without some warning message flashing at you when reviewing your nuget package manager in my opinion.  This is surely out of your control and a Microsoft issue to me now.  They made the solution work with Server but not Client, why?!  It just is confusing and should not even raise an eyebrow if not a problem.
